client/verta/verta/runtime.py

```diff
@@ -0,0 +1,237 @@
+       :emphasize-lines: 11,14
+
+        from verta import runtime
+
+        # Sample model code:
+        class MyModel(VertaModelBase):
+            def __init__(self, artifacts):
+                pass
+
+            @verify.io
+            def predict(self, data):
+                logs = {'some_stuff': 'I_care_about'}
+                runtime.log('my_logging_key', logs)
+                prediction_output = 'some_output'
+                more_logs = ['a', 'list', 'of', 'things']
+                runtime.log('things', more_logs)
+                return prediction_output
+
+    """
+    if _get_validate_flag():
+        _validate_json(value)
+    _validate_s3(key)
+    local_logs: Dict[str, Any] = _get_thread_logs()
+    local_logs.update({key: value})
+
+
+class context:
+    """
+    Context manager for aggregating key-value pairs into a custom log entry.
+    It should not be necessary to instantiate this class directly unless you
+    wish to export log entries manually to a resource other than Amazon S3.
+
+    For all models deployed in Verta with active endpoints, the :meth:`~verta.registry.VertaModelBase.predict`
+    function of the model is wrapped inside an object of this class by default.
+    Logs collected inside the `predict()` function are exported to S3.
+
+    Parameters
+    ----------
+    validate : bool, default False
+        If true, each individual call to :meth:`~verta.runtime.log`  will
+        verify that the value provided is JSON serializable.
+
+    Raises
+    ------
+    RuntimeError
+        If logging context has been added via :meth:`~verta.runtime.log()`
+        outside the scope of the model's :meth:`~verta.registry.VertaModelBase.predict`
+        method, or outside the scope of this context manager.
+
+    Examples
+    --------
+
+    .. code-block:: python
+       :emphasize-lines: 4,8,11
+
+        import json
+        from verta import runtime
+
+        with runtime.context() as ctx:
+            runtime.log('labels', {'model_type': 'scikit_learn'})
+            output = 'this_is_output'
+            runtime.log('output', output)
+            print json.dumps(ctx.logs())
+
+        # After exiting the context manager:
+        final_log_entry = json.dumps(ctx.logs())
+
+        # Output (both) = {"labels": {"model_type": "scikit_learn"}, "output": "this_is_output"}
+
+    """
+    def __init__(self, validate: Optional[bool] = False):
+        self._validate = validate
+        self._logs_dict = dict()
+
+    def __enter__(self):
+        """
+        Ensure an empty logging context to start and set validation flag.
+        """
+        if bool(_get_thread_logs()):  # If not an empty dict.
+            raise RuntimeError(
+                " cannot overwrite prior logging context. Please ensure all calls"
+                " to runtime.log() are made within the predict() method of your"
+                " model, or inside the scope of an instance of verta.runtime.context()."
```

Comment:
To put it another way, if we wait until `with runtime.context():` to tell the user that they called `runtime.log()` in the wrong place, it's already too late! There's no public way for them to recover/access that log they made, and there are no public functions for them to fix their thread state. We should raise that error in `runtime.log()` to prevent the problem from happening in the first place.